Simplify boolean return logic

nickdrozd opened the pull request:
I was poking around in the code and I noticed that there are a lot of bare boolean returns. For example:

```python
        if self.subtype_context.keep_erased_types:
            return False
        return True
```

This can be simplified to:

```python
        return not self.subtype_context.keep_erased_types
```

That's a pretty simple case. Other cases are more complicated, but similar in spirit.

This is a refactoring PR that should not change any behavior. By my count it results in a net reduction of 32 `return` statements.

If this sort of change is welcome, I can keep going.

hauntsaninja commented on mypy/subtypes.py:
```diff
@@ -403,48 +403,43 @@ def visit_any(self, left: AnyType) -> bool:
-        if self.subtype_context.keep_erased_types:
-            return False
-        return True
+        return not self.subtype_context.keep_erased_types
```
keep

hauntsaninja commented on mypy/subtypes.py:
```diff
@@ -403,48 +403,43 @@ def visit_any(self, left: AnyType) -> bool:
+            # can by a subtype of it, even with an `Any` fallback.
+            # This special case is needed to treat descriptors in classes with
+            # dynamic base classes correctly, see #5456.
+            return not isinstance(self.right, NoneType)
```
keep

hauntsaninja commented on mypy/subtypes.py:
```diff
@@ -513,11 +508,7 @@ def check_mixed(
-                            ):
-                                return False
-                            return True
+                            return all(is_equivalent(l, unpacked_type.args[0]) for l in compare_to)
```
keep

hauntsaninja commented on mypy/subtypes.py:
```diff
@@ -735,15 +716,12 @@ def visit_tuple_type(self, left: TupleType) -> bool:
-            for l, r in zip(left.items, right.items):
-                if not self._is_subtype(l, r):
-                    return False
+            if any(not self._is_subtype(l, r) for l, r in zip(left.items, right.items)):
```
keep

hauntsaninja commented on mypy/subtypes.py:
```diff
@@ -752,9 +730,7 @@ def visit_tuple_type(self, left: TupleType) -> bool:
-            if not self._is_subtype(lfallback, rfallback):
-                return False
-            return True
+            return self._is_subtype(lfallback, rfallback)
```
keep

hauntsaninja commented on mypy/subtypes.py:
```diff
@@ -811,10 +787,7 @@ def visit_overloaded(self, left: Overloaded) -> bool:
-                if self._is_subtype(item, right):
-                    return True
-            return False
+            return any(self._is_subtype(item, right) for item in left.items)
```
keep

hauntsaninja commented on mypy/subtypes.py:
```diff
@@ -988,9 +961,8 @@ def f(self) -> A: ...
-    for (l, r) in reversed(assuming):
-        if l == left and r == right:
-            return True
+    if any(l == left and r == right for (l, r) in reversed(assuming)):
```
keep

hauntsaninja commented on mypy/subtypes.py:
```diff
@@ -1376,8 +1348,7 @@ def g(x: int) -> int: ...
             return False
-        else:
-            left = unified
+        left = unified
```
keep

hauntsaninja commented on mypy/types.py:
```diff
@@ -437,14 +437,12 @@ def __repr__(self) -> str:
-            )
-        else:
-            return False
+        return (
```
keep

hauntsaninja commented on mypy/types.py:
```diff
@@ -910,9 +908,7 @@ def __hash__(self) -> int:
-        if not isinstance(other, TypeList):
-            return False
-        return self.items == other.items
+        return isinstance(other, TypeList) and self.items == other.items
```
keep

hauntsaninja commented on mypy/types.py:
```diff
@@ -2263,16 +2261,19 @@ def __hash__(self) -> int:
-                    return False
-            return self.fallback == other.fallback and self.required_keys == other.required_keys
-        else:
+        if not isinstance(other, TypedDictType):
```
keep

hauntsaninja commented on mypy/types.py:
```diff
@@ -3346,21 +3344,19 @@ def is_literal_type(typ: ProperType, fallback_fullname: str, value: LiteralValue
-    if typ.fallback.type.fullname != fallback_fullname:
-        return False
-    return typ.value == value
+    return (
```
keep

hauntsaninja reviewed (changes requested):
Thanks, but several of these make the code less readable. In general, some of these sequences of if conditions are pretty subtle and it's much easier to read them one at a time than to have them compressed into a single statement. I've commented on the ones I think are worthwhile, if you want to pursue this PR, please remove the others :-)

nickdrozd commented:
Okay, I think I got all of them.

hauntsaninja commented on mypy/types.py:
```diff
@@ -2263,16 +2259,19 @@ def __hash__(self) -> int:
+        return (
+            frozenset(self.items.keys()) == frozenset(other.items.keys())
+            and all(
+                not left_item_type == right_item_type
```
This needs fixing :-)

nickdrozd replied:
Fixed (I hope)

nickdrozd commented:
I fixed the existing changes and added a few more. All the new changes are pretty small, and they can be kept or not.

In the spirit of simplified boolean logic, it would be easier if you mark the ones to reject, rather than the ones to keep :smile:

hauntsaninja commented on mypy/checker.py:
```diff
@@ -6152,13 +6147,12 @@ def handle_partial_var_type(
+        if not var.info:
+            return False
+        return (
+            any(base.get(var.name) is not None for base in var.info.mro[1:])
+            or var.info.fallback_to_any
```
```suggestion
            var.info.fallback_to_any
            or any(base.get(var.name) is not None for base in var.info.mro[1:])
```

nickdrozd replied:
This seems like a reasonable change, although unlike the other ones it isn't strictly equivalent.

hauntsaninja commented on mypy/meet.py:
```diff
@@ -535,9 +535,7 @@ def are_typed_dicts_overlapping(
-        if key not in right.items:
-            return False
-        if not is_overlapping_types(
+        if key not in right.items or not is_overlapping_types(
```
revert

hauntsaninja commented on mypy/meet.py:
```diff
@@ -547,9 +545,7 @@ def are_typed_dicts_overlapping(
-        if key not in left.items:
-            return False
-        if not is_overlapping_types(
+        if key not in left.items or not is_overlapping_types(
```
revert

hauntsaninja commented on mypy/checker.py:
```diff
@@ -6348,14 +6342,11 @@ def conditional_types_with_intersection(
-            if node.is_property and not node.is_settable_property:
-                return False
-            return True
+            return not node.is_property or node.is_settable_property
```
revert

hauntsaninja commented on mypy/stubgen.py:
```diff
@@ -1489,32 +1486,28 @@ def find_module_paths_using_imports(
-            ".__main__",
-            ".conftest",  # Used by pytest
-            ".setup",  # Typically an install script
+    return (
```
revert

hauntsaninja commented on mypyc/ir/class_ir.py:
```diff
@@ -248,15 +248,12 @@ def is_method_final(self, name: str) -> bool:
-                    return False
-            return True
-        else:
+        if not self.has_method(name):
```
revert

hauntsaninja commented on mypyc/ir/class_ir.py:
```diff
@@ -265,15 +262,10 @@ def has_attr(self, name: str) -> bool:
-        if self.is_deletable(name):
-            return False
-        return name in self._always_initialized_attrs
+        return not self.is_deletable(name) and name in self._always_initialized_attrs
```
revert

hauntsaninja commented on mypyc/irbuild/ast_helpers.py:
```diff
@@ -115,6 +115,4 @@ def is_borrow_friendly_expr(self: IRBuilder, expr: Expression) -> bool:
-    if isinstance(expr, MemberExpr) and self.is_native_attr_ref(expr):
-        return True
-    return False
+    return isinstance(expr, MemberExpr) and self.is_native_attr_ref(expr)
```
revert

hauntsaninja reviewed (commented):
Thanks, reviewed the new changes. Please don't add any others in this PR :-)

hauntsaninja reviewed (approved):
Thank you!

nickdrozd commented:
Thanks for reviewing!
